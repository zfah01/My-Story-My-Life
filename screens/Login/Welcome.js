import React, { useState, useEffect } from 'react';
import { Text, View, ImageBackground, ScrollView, TextInput, Image, Alert, TouchableOpacity } from 'react-native';
import Loading from '../../utils/Loading';
import * as LocalAuthentication from 'expo-local-authentication';
import { styles } from './Styles';


import { auth, db } from '../../firebase/firebase';

//Check if email is in a suitable format 
export const emailValid = (str) => {
    const validEmail = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return validEmail.test(str);
};

export default function Welcome({ navigation }) {
    
    //Initiate states 
    const [email, setEmail] = useState('');
    const [password, setPassword] = useState('');
    const [loading, setLoading] = useState(false);

    //Add 2FA, prompts the user to enter with their Local Authentication (Biometrics, Passcode etc)
    // if success directs user to HomeScreen else error message displayed

        const getEnrollment = async () => {
            const isEnrolled = await LocalAuthentication.isEnrolledAsync();
            if (isEnrolled) {
                const { success } = await LocalAuthentication.authenticateAsync({ promptMessage: 'Authenticate'});
                if (success) {
                    navigation.navigate('HomeScreen');
                } else {
                    Alert.alert('Uh Oh! Access Denied. Please try again.')
                    navigation.navigate('Welcome');
                    
                }
                
            }
        };



    const onLoginPress = () => {
        setLoading(true);
        //Checks if email and password exists if it does it directs user to Home 

        if (!emailValid(email)) {
            setLoading(false);
            alert('Please ensure your email is valid.');
            return;
         }
         
        auth.
        signInWithEmailAndPassword(email, password)
            .then((response) => {
                const uid = response.user.uid;
                const usersRef = db.collection('users');
                usersRef
                    .doc(uid)
                    .get()
                    .then(firestoreDocument => {
                        if (!firestoreDocument.exists) {
                            alert('User does not exist. Please check your credentials.');
                            return;
                        }
                        setLoading(false);
                        getEnrollment();
                    })
                    .catch(error => {
                        setLoading(false);
                        alert(error);
                    });
            })
            .catch(error => {
                setLoading(false);
                alert(error);
            });
    };


    if (loading) {
        return (
            <Loading loading={loading} />
        );
    } else {
        return (
                <View style={{backgroundColor: '#AFEEEE', width: '100%', height: '100%'}}>
                <View style={styles.intro}>
                <Text style={styles.heading}>Welcome to My Story, My Life</Text>
                <Text>Log into your account below:</Text>
            </View><ScrollView>
                    <View style={styles.mainView}>
                        <Image style={styles.logo} source={require('../../assets/logo.png')} />

                        <TextInput
                            style={styles.textInput}
                            placeholder='enter your email address...'
                            placeholderTextColor='#aaaaaa'
                            onChangeText={(textEmail) => setEmail(textEmail)}
                            value={email}
                            autoCapitalize='none'
                            testID='inputEmail' />

                        <TextInput
                            style={styles.textInput}
                            secureTextEntry
                            placeholder='enter your password...'
                            placeholderTextColor='#aaaaaa'
                            onChangeText={(textPass) => setPassword(textPass)}
                            value={password}
                            autoCapitalize='none'
                            testID='inputPassword' />

                        <TouchableOpacity
                            style={styles.buttonLogin}
                            onPress={() => onLoginPress()}
                            testID='buttonLogin'
                            accessibilityLabel='Log In button'>
                            <Text style={styles.loginText}> Log In </Text>
                        </TouchableOpacity>

                        <View style={styles.footer}>
                            <Text style={styles.accountText}>Don&apos;t have an account yet?<Text testID='signUpLink' onPress={() => navigation.navigate('SignUp')} style={styles.createLink}> Sign up</Text> </Text>
                        </View>
                    </View>
                </ScrollView>
                </View>
        );
    }
}