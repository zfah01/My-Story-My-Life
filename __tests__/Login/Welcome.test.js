import 'react-native';
import React from 'react';
import Welcome, { emailValid } from '../../screens/Login/Welcome';

import renderer from 'react-test-renderer';

import { cleanup, render, fireEvent } from '@testing-library/react-native';

afterEach(cleanup);

describe('Testing components and functions of Welcome', () => {

  const fakeNavigation = {
    navigate: jest.fn(),
  };

  it('Checks to see if the Welcome page renders', () => {
    const tree = renderer.create(<Welcome />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Test for displaying greetings message', () => {
    const title = 'Welcome to My Story, My Life';
    const { getByText } = render(<Welcome />);
    const foundTitleText = getByText(title);

    expect(foundTitleText.props.children).toEqual(title);
  });

  it('Find button from testID', () => {
    const { getByTestId } = render(<Welcome />);
    const testIdName = 'buttonLogin';
    const buttonID = getByTestId(testIdName);

    expect(buttonID).toBeTruthy();
  });

  it('Test to see if email and password inputs receive data', async () => {
    const { getByTestId } = render(<Welcome />);
    const testIdEmail = 'inputEmail';
    const testIdPassword = 'inputPassword';
    const inputEmail = getByTestId(testIdEmail);
    const inputPassword = getByTestId(testIdPassword);

    fireEvent.changeText(inputEmail, 'test@example.com');
    fireEvent.changeText(inputPassword, 'password');

    expect(inputEmail.props.value).toEqual('test@example.com');
    expect(inputPassword.props.value).toEqual('password');
  });

  it('Checks that an email is valid', () => {
    const emailBad = 'hello';
    const emailGood = 'test@test.com';
    expect(emailValid(emailBad)).toBeFalsy();
    expect(emailValid(emailGood)).toBeTruthy();
  });

  it('Testing Sign Up page navigation', () => {
    const idSignUp = 'signUpLink';
    const { getByTestId } = render(<Welcome navigation={fakeNavigation} />);

    fireEvent(getByTestId(idSignUp), 'press');
    expect(fakeNavigation.navigate).toBeCalledWith('SignUp');
  });


});