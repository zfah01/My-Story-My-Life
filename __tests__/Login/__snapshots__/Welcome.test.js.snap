// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing components and functions of Welcome Checks to see if the Welcome page renders 1`] = `
Array [
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "paddingTop": 70,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 25,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Welcome to My Story, My Life
    </Text>
    <Text>
      Log into your account below:
    </Text>
  </View>,
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 25,
            "justifyContent": "center",
            "margin": 20,
            "padding": 10,
          }
        }
      >
        <Image
          source={1}
          style={
            Object {
              "alignSelf": "center",
              "height": 100,
              "marginBottom": 10,
              "marginTop": 20,
              "width": 134,
            }
          }
        />
        <TextInput
          autoCapitalize="none"
          onChangeText={[Function]}
          placeholder="enter your email address..."
          placeholderTextColor="#aaaaaa"
          style={
            Object {
              "backgroundColor": "#f0efed",
              "borderRadius": 5,
              "height": 50,
              "marginBottom": 10,
              "marginLeft": 20,
              "marginRight": 20,
              "marginTop": 10,
              "overflow": "hidden",
              "paddingLeft": 10,
            }
          }
          testID="inputEmail"
          value=""
        />
        <TextInput
          autoCapitalize="none"
          onChangeText={[Function]}
          placeholder="enter your password..."
          placeholderTextColor="#aaaaaa"
          secureTextEntry={true}
          style={
            Object {
              "backgroundColor": "#f0efed",
              "borderRadius": 5,
              "height": 50,
              "marginBottom": 10,
              "marginLeft": 20,
              "marginRight": 20,
              "marginTop": 10,
              "overflow": "hidden",
              "paddingLeft": 10,
            }
          }
          testID="inputPassword"
          value=""
        />
        <View
          accessibilityLabel="Log In button"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "#00e676",
              "borderRadius": 15,
              "justifyContent": "center",
              "marginTop": 20,
              "opacity": 1,
              "padding": 15,
              "width": 150,
            }
          }
          testID="buttonLogin"
        >
          <Text
            style={
              Object {
                "color": "#000000",
                "fontSize": 18,
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
             Log In 
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "marginTop": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 16,
              }
            }
          >
            Don't have an account yet?
            <Text
              onPress={[Function]}
              style={
                Object {
                  "color": "#6db8bf",
                  "fontSize": 16,
                  "fontWeight": "bold",
                }
              }
              testID="signUpLink"
            >
               Sign up
            </Text>
             
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>,
]
`;
